### PR TITLE
Prevent a reinstall

### PR DESCRIPTION
**Pull request by rullzer**

### CODE REVIEW

File: core/templates/installation_forbidden.php [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<div class="error">
+	<h2><?php p($l->t('Error')) ?></h2>
+	<p>
+		<?php p($l->t('It looks like you are trying to reinstall your Nextcloud. However the file CAN_INSTALL is missing from your config directory. Please create file CAN_CREATE in your config folder to continue.')) ?>
```
Comment (nickvergessen) [inline core/templates/installation_forbidden.php]:
```suggestion
		<?php p($l->t('It looks like you are trying to reinstall your Nextcloud. However the file CAN_INSTALL is missing from your config directory. Please create the file CAN_INSTALL in your config folder to continue.')) ?>
```

**Comment by MorrisJobke** [comment]
> [ ] Maybe a check in the admin settings?

for what? if the file is present? but the updater should remove it, right?

**Comment by nickvergessen** [comment]
I vote for a bypass on update channel `git`.
Otherwise this will be a pain for developers.

You always remove the file and commit it by accident and you need to recreate it, in order to be able to reinstall

**Comment by rullzer** [comment]
> I vote for a bypass on update channel `git`.
> Otherwise this will be a pain for developers.
> 
> You always remove the file and commit it by accident and you need to recreate it, in order to be able to reinstall

yes that was also my idea :+1:

**Comment by rullzer** [comment]
Done I think

**Review by nickvergessen — APPROVED** [review]

File: core/Controller/SetupController.php [hunk 2]
```diff
@@ -101,6 +110,11 @@ public function finishSetup() {
 			unlink($this->autoConfigFile);
 		}
 		\OC::$server->getIntegrityCodeChecker()->runInstanceVerification();
+
+		if (\OC_Util::getChannel() !== 'git' && is_file(\OC::$configDir.'/CAN_INSTALL')) {
+			unlink(\OC::$configDir.'/CAN_INSTALL');
```
Comment (ChristophWurst) [inline core/Controller/SetupController.php]: check return value to throw an exception in case of a failure?

File: lib/private/Updater.php [hunk 3]
```diff
@@ -111,6 +111,11 @@ public function upgrade() {
 			$this->emit('\OC\Updater', 'maintenanceEnabled');
 		}
 
+		// Clear CAN_INSTALL file if not on git
+		if (\OC_Util::getChannel() !== 'git') {
```
Comment (ChristophWurst) [inline lib/private/Updater.php]: would `is_file` make sense here as well?

File: lib/private/Updater.php [hunk 4]
```diff
@@ -111,6 +111,11 @@ public function upgrade() {
...
+			unlink(\OC::$configDir . '/CAN_INSTALL');
```
Comment (ChristophWurst) [inline lib/private/Updater.php]: error handling would be great like above

Comment (rullzer) [reply lib/private/Updater.php]: Logs an error now.

File: lib/private/Updater.php (hunk 3 above)
Comment (rullzer) [reply lib/private/Updater.php]: sure

File: core/Controller/SetupController.php (hunk 2 above)
Comment (rullzer) [reply core/Controller/SetupController.php]: fair enough

**Review by ChristophWurst — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]

**Comment by nickvergessen** [comment]
Was merged before stable16 right?

**Comment by ChristophWurst** [comment]
Yes: https://github.com/nextcloud/server/blob/stable16/config/CAN_INSTALL

**Pull request merged**